llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -463,6 +463,15 @@ std::string saveModuleProperties(module_split::ModuleDesc &MD,
     if (HasLargeGRF)
       PropSet[PropSetRegTy::SYCL_MISC_PROP].insert({"isLargeGRF", true});
   }
+  {
+    // check for auto GRF property
+    bool HasAutoGRF = llvm::any_of(MD.entries(), [](const Function *F) {
```

Comment:
Another option would be to not add a new property,  remove the bool  `isLargeGRF` and add a new int property `RegisterAllocMode` there the values match the `RegisterAllocMode` metadata, and check for that in the runtime.
 
 There's a small tradeoff between readbility and the extra property, so let me know if you have an opinion on which to do.

Reply:
There is a `TODO` item in `kernel_properties.hpp` to completely rewrite this `large-grf` property from what started as a temporary API (9a55da536b1455cfa5257ded077f2b6322332357) to a solution, which is supposed to be generic for all properties:

https://github.com/intel/llvm/blob/e6a30753164e47932a6e9dfc81ebf68d2cd1051f/sycl/include/sycl/ext/intel/experimental/kernel_properties.hpp#L67-L69

`sycl_ext_oneapi_kernel_properties` is marked as `experimental` since Nov 10th, 2022 (e39e8eba641d9b155392964b0ec0eddd7c91266d). Perhaps it is time to do this refactoring? 

The reason I'm asking is because design of user-facing part of the extension can influence underlying implementation design (at least make some ways more preferrable to align between components).

My preference here would be to have a single `RegisterAllocMode` property, because I think that with `sycl_ext_oneapi_kernel_properties` we won't have several properties at SYCL level anymore. I imagine that instead we would have a single `intel::register_allocation_mode` property which would accept values like `large` or `auto`.

We have an [ABI-breaking window] open for the next month, so we could use this opportunity to perform a complete switchover from old device image properties format to a new one in a single patch, without going through deprecation path first. (However, since `large-grf` is experimental anyway, we are allowed to break it a bit more often than non-experimental stuff).

But before we go to the actual implementation, I have a bigger concern: where is the language specification for this feature? I wasn't able to find one neither by looking through [`extensions`](https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions) folder, nor by looking at commits which [modified](https://github.com/intel/llvm/commits/e6a30753164e47932a6e9dfc81ebf68d2cd1051f/sycl/include/sycl/ext/intel/experimental/kernel_properties.hpp) `kernel_properties.hpp` file.

My main question is: why do we want `auto` to be explicitly set by user? Why can't it be done automatically by a backend compiler with an option for user to explicitly lock a kernel for a certain register allocation mode?

What is the default register allocation mode right now? Can user specify current default explicitly to be sure that their code won't suddenly behave differently if we change the default?

Reply:
> If you are okay with idea overall (adding per-kernel auto support)

I don't have objections against adding `auto` mode: the extension is experimental and we are clearly trying to gather hands-on experience with the feature and customer feedback from it, before generalizing it and proposing it as a formal extension. I also understand that changing the default will likely cause a lot of side effects and we may not be ready for that yet.

Therefore, I'm fine with implementing the suggested language design, but I do have a couple of comments about its details:
1. Let's use current ABI-breaking window to change underlying format of device image properties used to represent `grf`-related flags. I suggest we introduce some single integer property, which would represent different values we need. This would allow us to later add more `grf` modes without changing properties at all
2. (lower priority) Let's migrate user-facing interface to [`sycl_ext_oneapi_kernel_properties`](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc). Even though we don't have a formal spec for `large-grf`, I assume that we would want to re-use other interfaces we proposed so that property is unified with other properties.

Reply:
Thanks, I'll try to do both in this PR, let me implement that